File: app/controllers/webhooks_controller.rb
class WebhooksController < ApplicationController
  skip_before_action :verify_authenticity_token
  before_action :verify_twilio_signature, only: %i[twilio_sms_status twilio_voice_status twilio_trust_hub]

  # Twilio SMS Status Webhook
  def twilio_sms_status
    # Use find_or_initialize_by to detect if it's a new record
    webhook = Webhook.find_or_initialize_by(
      source: 'twilio_sms',
      external_id: params[:MessageSid]
    )

    if webhook.new_record?
      webhook.event_type = 'sms_status'
      webhook.payload = webhook_params
      webhook.headers = extract_headers
      webhook.status = 'pending'
      webhook.received_at = Time.current

      if webhook.save
        WebhookProcessorJob.perform_later(webhook.id)
      else
        Rails.logger.error("Failed to create SMS webhook: #{webhook.errors.full_messages.join(', ')}")
      end
    elsif webhook.status != 'pending'
      Rails.logger.info("Duplicate SMS webhook rejected (already #{webhook.status}): #{params[:MessageSid]}")
    else
      # Duplicate request for pending webhook - ignore (idempotent)
      Rails.logger.info("Duplicate SMS webhook ignored (pending): #{params[:MessageSid]}")
    end

    head :ok
  rescue ActiveRecord::RecordNotUnique
    Rails.logger.warn("Duplicate SMS webhook (race condition): #{params[:MessageSid]}")
    head :ok
  rescue StandardError => e
    Rails.logger.error("SMS webhook error: #{e.class} - #{e.message}")
    head :ok
  end

  # Twilio Voice Status Webhook
  def twilio_voice_status
    webhook = Webhook.find_or_initialize_by(
      source: 'twilio_voice',
      external_id: params[:CallSid]
    )

    if webhook.new_record?
      webhook.event_type = 'voice_status'
      webhook.payload = webhook_params
      webhook.headers = extract_headers
      webhook.status = 'pending'
      webhook.received_at = Time.current

      if webhook.save
        WebhookProcessorJob.perform_later(webhook.id)
      else
        Rails.logger.error("Failed to create Voice webhook: #{webhook.errors.full_messages.join(', ')}")
      end
    elsif webhook.status != 'pending'
      Rails.logger.info("Duplicate Voice webhook rejected (already #{webhook.status}): #{params[:CallSid]}")
    else
      Rails.logger.info("Duplicate Voice webhook ignored (pending): #{params[:CallSid]}")
    end

    head :ok
  rescue ActiveRecord::RecordNotUnique
    Rails.logger.warn("Duplicate Voice webhook (race condition): #{params[:CallSid]}")
    head :ok
  rescue StandardError => e
    Rails.logger.error("Voice webhook error: #{e.class} - #{e.message}")
    head :ok
  end

  # Twilio Trust Hub Status Webhook
  def twilio_trust_hub
    webhook = Webhook.find_or_initialize_by(
      source: 'twilio_trust_hub',
      external_id: params[:CustomerProfileSid]
    )

    if webhook.new_record?
      webhook.event_type = params[:StatusCallbackEvent] || 'status_update'
      webhook.payload = webhook_params
      webhook.headers = extract_headers
      webhook.status = 'pending'
      webhook.received_at = Time.current

      if webhook.save
        WebhookProcessorJob.perform_later(webhook.id)
      else
        Rails.logger.error("Failed to create Trust Hub webhook: #{webhook.errors.full_messages.join(', ')}")
      end
    elsif webhook.status != 'pending'
      Rails.logger.info("Duplicate Trust Hub webhook rejected (already #{webhook.status}): #{params[:CustomerProfileSid]}")
    else
      Rails.logger.info("Duplicate Trust Hub webhook ignored (pending): #{params[:CustomerProfileSid]}")
    end

    head :ok
  rescue ActiveRecord::RecordNotUnique
    Rails.logger.warn("Duplicate Trust Hub webhook (race condition): #{params[:CustomerProfileSid]}")
    head :ok
  rescue StandardError => e
    Rails.logger.error("Trust Hub webhook error: #{e.class} - #{e.message}")
    head :ok
  end

  # Generic webhook endpoint (requires API key authentication)
  # Usage: POST /webhooks/receive with Authorization: Bearer <api_token>
  def generic
    # Verify API key authentication
    unless authenticate_webhook_api_key
      render json: { success: false, error: 'Unauthorized: Invalid or missing API key' }, status: :unauthorized
      return
    end

    webhook = Webhook.create(
      source: params[:source] || 'unknown',
      event_type: params[:event_type] || 'unknown',
      external_id: params[:external_id],
      payload: webhook_params,
      headers: extract_headers,
      status: 'pending',
      received_at: Time.current
    )

    if webhook.persisted?
      WebhookProcessorJob.perform_later(webhook.id)
      render json: { success: true, webhook_id: webhook.id }
    else
      render json: { success: false, errors: webhook.errors.full_messages }, status: :unprocessable_entity
    end
  rescue StandardError => e
    Rails.logger.error("Generic webhook error: #{e.class} - #{e.message}")
    render json: { success: false, error: e.message }, status: :internal_server_error
  end

  private

  # Authenticate generic webhook using API key from Authorization header
  def authenticate_webhook_api_key
    auth_header = request.headers['Authorization']
    return false unless auth_header.present?

    # Extract Bearer token
    token = auth_header.split(' ').last
    return false unless token.present?

    # Validate against admin user API tokens
    AdminUser.exists?(api_token: token)
  end

  def verify_twilio_signature
    # Verify webhook is from Twilio using signature validation
    app_creds = defined?(AppConfig) ? AppConfig.twilio_credentials : nil
    auth_token = app_creds&.dig(:auth_token) || TwilioCredential.current&.auth_token
    return head :forbidden unless auth_token.present?

    validator = Twilio::Security::RequestValidator.new(auth_token)
    signature = request.headers['HTTP_X_TWILIO_SIGNATURE']
    url = request.original_url

    # Combine POST parameters and query parameters for signature validation
    # Twilio may send data in either or both locations
    params_for_validation = request.POST.merge(request.query_parameters)

    unless validator.validate(url, params_for_validation, signature)
      Rails.logger.warn "Invalid Twilio signature for webhook: #{request.path}"
      head :forbidden
    end
  rescue ArgumentError, TypeError => e
    # Handle invalid auth token or malformed signature
    Rails.logger.error "Signature verification error: #{e.class} - #{e.message}"
    head :forbidden
  rescue StandardError => e
    # Unexpected errors during validation
    Rails.logger.error "Unexpected signature verification error: #{e.class} - #{e.message}"
    Rails.logger.error e.backtrace.join("\n")
    head :forbidden
  end

  def webhook_params
    params.except(:controller, :action, :format).to_unsafe_h
  end

  def extract_headers
    {
      'User-Agent' => request.headers['HTTP_USER_AGENT'],
      'X-Twilio-Signature' => request.headers['HTTP_X_TWILIO_SIGNATURE'],
      'Content-Type' => request.headers['CONTENT_TYPE']
    }
  end
end

File: laravel-app/app/Http/Controllers/WebhooksController.php
<?php

namespace App\Http\Controllers;

use App\Models\Webhook;
use App\Models\AdminUser;
use App\Models\TwilioCredential;
use App\Jobs\WebhookProcessorJob;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;
use Twilio\Security\RequestValidator;

class WebhooksController extends Controller
{
    /**
     * Twilio SMS Status Webhook
     */
    public function twilioSmsStatus(Request $request)
    {
        try {
            // Verify Twilio signature
            if (!$this->verifyTwilioSignature($request)) {
                Log::warning('Invalid Twilio signature for webhook: ' . $request->path());
                return response('Forbidden', 403);
            }

            // Atomic upsert to prevent race conditions
            $data = [
                'source' => 'twilio_sms',
                'external_id' => $request->input('MessageSid'),
                'event_type' => 'sms_status',
                'payload' => json_encode($this->getWebhookParams($request)),
                'headers' => json_encode($this->extractHeaders($request)),
                'status' => 'pending',
                'received_at' => now(),
                'created_at' => now(),
                'updated_at' => now(),
            ];

            // Use INSERT ... ON DUPLICATE KEY UPDATE
            $webhook = Webhook::updateOrCreate(
                [
                    'source' => 'twilio_sms',
                    'external_id' => $request->input('MessageSid'),
                ],
                $data
            );

            // Check if this was a new insert (wasRecentlyCreated)
            if ($webhook->wasRecentlyCreated) {
                WebhookProcessorJob::dispatch($webhook->id);
            } else {
                Log::info("Duplicate SMS webhook ignored: {$request->input('MessageSid')}");
            }

            return response('', 200);
        } catch (\Exception $e) {
            Log::error("SMS webhook error: " . get_class($e) . " - " . $e->getMessage());
            return response('', 200);
        }
    }

    /**
     * Twilio Voice Status Webhook
     */
    public function twilioVoiceStatus(Request $request)
    {
        try {
            // Verify Twilio signature
            if (!$this->verifyTwilioSignature($request)) {
                Log::warning('Invalid Twilio signature for webhook: ' . $request->path());
                return response('Forbidden', 403);
            }

            // Atomic upsert to prevent race conditions
            $data = [
                'source' => 'twilio_voice',
                'external_id' => $request->input('CallSid'),
                'event_type' => 'voice_status',
                'payload' => json_encode($this->getWebhookParams($request)),
                'headers' => json_encode($this->extractHeaders($request)),
                'status' => 'pending',
                'received_at' => now(),
                'created_at' => now(),
                'updated_at' => now(),
            ];

            $webhook = Webhook::updateOrCreate(
                [
                    'source' => 'twilio_voice',
                    'external_id' => $request->input('CallSid'),
                ],
                $data
            );

            if ($webhook->wasRecentlyCreated) {
                WebhookProcessorJob::dispatch($webhook->id);
            } else {
                Log::info("Duplicate Voice webhook ignored: {$request->input('CallSid')}");
            }

            return response('', 200);
        } catch (\Exception $e) {
            Log::error("Voice webhook error: " . get_class($e) . " - " . $e->getMessage());
            return response('', 200);
        }
    }

    /**
     * Twilio Trust Hub Status Webhook
     */
    public function twilioTrustHub(Request $request)
    {
        try {
            // Verify Twilio signature
            if (!$this->verifyTwilioSignature($request)) {
                Log::warning('Invalid Twilio signature for webhook: ' . $request->path());
                return response('Forbidden', 403);
            }

            // Atomic upsert to prevent race conditions
            $data = [
                'source' => 'twilio_trust_hub',
                'external_id' => $request->input('CustomerProfileSid'),
                'event_type' => $request->input('StatusCallbackEvent', 'status_update'),
                'payload' => json_encode($this->getWebhookParams($request)),
                'headers' => json_encode($this->extractHeaders($request)),
                'status' => 'pending',
                'received_at' => now(),
                'created_at' => now(),
                'updated_at' => now(),
            ];

            $webhook = Webhook::updateOrCreate(
                [
                    'source' => 'twilio_trust_hub',
                    'external_id' => $request->input('CustomerProfileSid'),
                ],
                $data
            );

            if ($webhook->wasRecentlyCreated) {
                WebhookProcessorJob::dispatch($webhook->id);
            } else {
                Log::info("Duplicate Trust Hub webhook ignored: {$request->input('CustomerProfileSid')}");
            }

            return response('', 200);
        } catch (\Exception $e) {
            Log::error("Trust Hub webhook error: " . get_class($e) . " - " . $e->getMessage());
            return response('', 200);
        }
    }

    /**
     * Generic webhook endpoint (requires API key authentication)
     * Usage: POST /webhooks/generic with Authorization: Bearer <api_token>
     */
    public function generic(Request $request)
    {
        try {
            // Verify API key authentication
            if (!$this->authenticateWebhookApiKey($request)) {
                return response()->json([
                    'success' => false,
                    'error' => 'Unauthorized: Invalid or missing API key'
                ], 401);
            }

            $webhook = Webhook::create([
                'source' => $request->input('source', 'unknown'),
                'event_type' => $request->input('event_type', 'unknown'),
                'external_id' => $request->input('external_id'),
                'payload' => $this->getWebhookParams($request),
                'headers' => $this->extractHeaders($request),
                'status' => 'pending',
                'received_at' => now(),
            ]);

            if ($webhook->exists) {
                WebhookProcessorJob::dispatch($webhook->id);
                return response()->json([
                    'success' => true,
                    'webhook_id' => $webhook->id
                ]);
            }

            return response()->json([
                'success' => false,
                'errors' => $webhook->getErrors()
            ], 422);
        } catch (\Exception $e) {
            Log::error("Generic webhook error: " . get_class($e) . " - " . $e->getMessage());
            return response()->json([
                'success' => false,
                'error' => $e->getMessage()
            ], 500);
        }
    }

    /**
     * Authenticate generic webhook using API key from Authorization header
     * Uses timing-safe comparison to prevent timing attacks
     */
    private function authenticateWebhookApiKey(Request $request): bool
    {
        $authHeader = $request->header('Authorization');
        if (!$authHeader) {
            return false;
        }

        // Extract Bearer token
        $parts = explode(' ', $authHeader);
        $token = end($parts);
        if (!$token) {
            return false;
        }

        // Timing-safe token validation to prevent brute-force via timing analysis
        $adminTokens = AdminUser::whereNotNull('api_token')->pluck('api_token');
        foreach ($adminTokens as $storedToken) {
            if (hash_equals($storedToken, $token)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Verify webhook is from Twilio using signature validation
     */
    private function verifyTwilioSignature(Request $request): bool
    {
        try {
            // Get auth token from TwilioCredential model or AppConfig
            $authToken = TwilioCredential::current()?->auth_token;
            if (!$authToken) {
                return false;
            }

            $validator = new RequestValidator($authToken);
            $signature = $request->header('X-Twilio-Signature');
            $url = $request->fullUrl();

            // Combine POST parameters and query parameters for signature validation
            $paramsForValidation = array_merge(
                $request->post(),
                $request->query()
            );

            return $validator->validate($signature, $url, $paramsForValidation);
        } catch (\ArgumentCountError | \TypeError $e) {
            // Handle invalid auth token or malformed signature
            Log::error("Signature verification error: " . get_class($e) . " - " . $e->getMessage());
            return false;
        } catch (\Exception $e) {
            // Unexpected errors during validation
            Log::error("Unexpected signature verification error: " . get_class($e) . " - " . $e->getMessage());
            Log::error($e->getTraceAsString());
            return false;
        }
    }

    /**
     * Extract all webhook parameters except controller/action
     */
    private function getWebhookParams(Request $request): array
    {
        $all = $request->all();
        unset($all['controller'], $all['action']);
        return $all;
    }

    /**
     * Extract relevant headers from the request
     */
    private function extractHeaders(Request $request): array
    {
        return [
            'User-Agent' => $request->header('User-Agent'),
            'X-Twilio-Signature' => $request->header('X-Twilio-Signature'),
            'Content-Type' => $request->header('Content-Type'),
        ];
    }
}
